docs/rfcs/cypher-based-hooks.md

```diff
@@ -0,0 +1,254 @@
+
+1. pre
+2. post
+
+```gql
+directive @pre(
+    cypher: String!
+    operations: [Operations!]! = [CREATE, READ, UPDATE, DELETE, DISCONNECT, CONNECT]
+) on OBJECT
+
+directive @post(
+    cypher: String!
+    operations: [Operations!]! = [CREATE, READ, UPDATE, DELETE, DISCONNECT, CONNECT]
+) on OBJECT
+```
+
+## Usage Examples
+
+### Audit Logs
+
+This is a real-world example! Most applications will at some point want to store a trail of events, and given that we have established that we don’t have a deterministic way of knowing if a node or relationship has been updated... The only logical and safe way would be to use a Cypher Hook. Given the schema below, there is a user and an audit node declared, with a relationship between them. The audit node is excluded from upserts and only created in the Hook.
+
+```gql
+type User
+    @post(
+        cypher: """
+        CREATE (this)-[:HAS_AUDIT]->(a:Audit)
+        SET a += {
+            id: randomUUID(),
+            createdAt: datetime(),
+            msg: 'User updated',
+            jwt: $auth.jwt.sub
+        }
+        """
+        operations: [UPDATE]
+    ) {
+    id: ID! @id
+    name: String!
+    email: String!
+    audits: [Audit!]! @relationship(type: "HAS_AUDIT", direction: OUT) @readonly
+}
+
+type Audit @exclude(operations: [CREATE, UPDATE, DELETE, CONNECT, DISCONNECT]) {
+    id: ID!
+    createdAt: DateTime!
+    msg: String!
+    jwt: String!
+}
+```
+
+Notice two things:
+
+1. The usage of `this` in the hook
+2. The usage of the `$jwt` and implies `$context` variables.
+
+Given that a user goes and preforms a GraphQL update mutation:
+
+```gql
+mutation {
+    updateUsers(where: { name: "cat" }, update: { email: "cat@cat.com" }) {
+        users {
+            name
+            email
+        }
+    }
+}
+```
+
+The cypher produced using APOC would be along the lines of:
+
+```gql
+CALL {
+    MATCH (u:User)
+    WHERE u.name = "cat"
+    SET u.email = "cat@cat.com"
+    CALL apoc.cypher.runFirstColumn("
+        CREATE (this)-[:HAS_AUDIT]->(a:Audit)
+        SET a += {
+            id: randomUUID(),
+            createdAt: datetime(),
+            msg: 'User updated',
+            jwt: $auth.jwt.sub
+        }
+    ", { this: u, $auth: auth }, false)
+    RETURN u
+}
+RETURN u { .name, .email } AS u
+```
+
+If we can get it to work without APOC even better:
+
+```gql
+CALL {
+    MATCH (u:User)
+    WHERE u.name = "cat"
+    SET u.email = "cat@cat.com"
+    CALL {
+        WITH u AS this
+        CREATE (this)-[:HAS_AUDIT]->(a:Audit)
+        SET a += {
+            id: randomUUID(),
+            createdAt: datetime(),
+            msg: 'User updated',
+            jwt: $auth.jwt.sub
+        }
+        RETURN count(*)
+    }
+    RETURN u
+}
+RETURN u { .name, .email } AS u
+```
+
+### Subscriptions
```

Comment:
Let's just completely remove this example from this RFC for brevity - we discussed this a long time ago and this just wouldn't work in a production environment.